File: brainstorming/design_studies/db_read/postgres_query.py
# design_studies/db_read/postgres_query.py
#
# Connect to Postgres and print rows from a `users` table.
# Companion to the Java, Rust (minimal/production), and
# Kāra (direct/injected) variants in this directory.
#
# Requires: pip install psycopg[binary]

import os
import sys

import psycopg


def main() -> None:
    url = os.environ.get("DATABASE_URL")
    if url is None:
        print("DATABASE_URL not set", file=sys.stderr)
        sys.exit(1)

    try:
        with psycopg.connect(url) as conn, conn.cursor() as cur:
            cur.execute("SELECT id, name, email FROM users ORDER BY id")
            for id_, name, email in cur:
                print(f"{id_}\t{name}\t{email}")
    except psycopg.Error as e:
        print(f"DB error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: brainstorming/design_studies/db_read/PostgresQuery.java
// design_studies/db_read/PostgresQuery.java
//
// Connect to Postgres and print rows from a `users` table.
// Companion to the Python, Rust (minimal/production), and
// Kāra (direct/injected) variants in this directory.
//
// Requires the postgresql JDBC driver on the classpath.

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class PostgresQuery {
    public static void main(String[] args) {
        String url = System.getenv("DATABASE_URL");
        if (url == null) {
            System.err.println("DATABASE_URL not set");
            System.exit(1);
        }

        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(
                 "SELECT id, name, email FROM users ORDER BY id");
             ResultSet rs = stmt.executeQuery()) {

            while (rs.next()) {
                long id = rs.getLong("id");
                String name = rs.getString("name");
                String email = rs.getString("email");
                System.out.println(id + "\t" + name + "\t" + email);
            }
        } catch (SQLException e) {
            System.err.println("DB error: " + e.getMessage());
            System.exit(1);
        }
    }
}
